transformations/lexical_counterfactual_generator/transformation.py

```diff
@@ -0,0 +1,111 @@
+from interfaces.SentencePairOperation import SentencePairOperation
+from tasks.TaskTypes import TaskType
+from typing import Tuple, List
+from initialize import spacy_nlp
+import spacy
+
+from transformations.back_translation import BackTranslation
+
+"""
+This generates counterfactuals using a simple rule based approach which would be 
+beneficial for paraphrase generation tasks.
+Assumes text pairs are labelled "1" for semantically similar and "0" for dissimilar.
+This is an example generator.
+"""
+
+
+def get_tokens_of_pos_type(sentence, types: list):
+    tokens = []
+    for token in sentence:
+        if token.pos_ in types:
+            tokens.append(token)
+    return tokens
+
+
+class LexicalCounterfactualGenerator(SentencePairOperation):
+    tasks = [
+        TaskType.PARAPHRASE_DETECTION
+    ]
+    languages = ["en"]
+    heavy = True
+    to_avoid = ["no", "not", "nowhere", "hardly", "barely", "nobody",
+                "none", "nothing", "neither", "nowhere", "never", "scarcely", "n't", "nt"]
+
+    def __init__(self, seed=0, max_outputs=2, pos_label="1", neg_label="0"):
+        super().__init__(seed, max_outputs=max_outputs)
+        self.nlp = spacy_nlp if spacy_nlp else spacy.load("en_core_web_sm")
+        self.pos_label = pos_label
+        self.neg_label = neg_label
+        self.bt = BackTranslation()
+
+    def negate_if_single_verb(self, sentence):
+        """
+        This only negates single verb sentences to be precise. This should be later extended to multiple verbs.
+        Eventually backtranslates the sentence to repair grammatical errors.
+        """
+        if "not " in sentence.split(" "):
+            return [sentence]
+        doc = self.nlp(sentence)
+        verbs = get_tokens_of_pos_type(doc, ["VERB"])
+        if len(verbs) == 1:
+            tokens = []
+            for token in doc:
+                if token.text in self.to_avoid:
+                    # do not generate if a negated word is present
+                    return [sentence]
+                if token.pos_ in ["VERB"]:
+                    tokens.append("not")
+                tokens.append(token.text)
+            basic_negated_sentence = " ".join(tokens)
+            return self.bt.generate(basic_negated_sentence)
+        return [sentence]
+
+    def generate(self, sentence1: str, sentence2: str, target: str) -> List[Tuple[str, str, str]]:
+        """
+        This implementation augments data for semantic similarity tasks.
+        This converts pairs which are positive (i.e. paraphrases-> label:1) to non-paraphrases (label:0).
+        """
+        examples = []
+        if target == self.pos_label:
+            for sentence1_n in self.negate_if_single_verb(sentence1):
```

Comment:
It seems that the `negate_if_single_verb` returns a list of only one sentence, then there's no need to do a for loop here.

Reply:
The `bt.generate()` function can generate multiple ones actually.. (line 61)